bson/bsonrw/json_scanner.go

```diff
@@ -202,13 +236,66 @@ func (js *jsonScanner) scanString() (*jsonToken, error) {
 					return nil, fmt.Errorf("invalid unicode sequence in JSON string: %s", us)
 				}
 
-				s := fmt.Sprintf(`\u%s`, us)
-				s, err = strconv.Unquote(strings.Replace(strconv.Quote(s), `\\u`, `\u`, 1))
-				if err != nil {
-					return nil, err
+				rn := getu4(us)
+
+				// If the rune we just decoded is the high or low value of a possible surrogate pair,
+				// try to decode the next sequence as the low value of a surrogate pair. We're
+				// expecting the next sequence to be another Unicode escape sequence (e.g. "\uDD1E"),
+				// but need to handle cases where the input is not a valid surrogate pair.
+				// For more context on unicode surrogate pairs, see:
+				// https://www.christianfscott.com/rust-chars-vs-go-runes/
+				// https://www.unicode.org/glossary/#high_surrogate_code_point
+				if utf16.IsSurrogate(rn) {
+					c, err = js.readNextByte()
+					if err != nil {
+						if err == io.EOF {
+							return nil, errors.New("end of input in JSON string")
+						}
+						return nil, err
+					}
+
+					// If the next value isn't the beginning of a backslash escape sequence, write the
+					// current rune and goto the beginning of the next char eval block.
+					if c != '\\' {
+						b.WriteRune(rn)
```

Comment:
To check my understanding. In this case, `b.WriteRune` will always write the replacement character, right?

`rn` is a surrogate sequence, but not part of a pair, so will necessarily be an invalid codepoint.

Reply:
I believe the response to https://github.com/mongodb/mongo-go-driver/pull/649/files#r624860862 answers this question.

> rn will never individually be a valid Unicode value

So this will always write a replacement character.

Reply:
I forgot to address this in my review, but you're correct. For consistency, I've updated the code to explicitly write the Unicode replacement character in all cases that the rune is a UTF-16 surrogate value but not part of a surrogate pair.